edfi_api_client/edfi_endpoint.py

```diff
@@ -1,711 +1,533 @@
-        Rows are returned as a generator.
-
-        :param page_size:
-        :param retry_on_failure:
-        :param max_retries:
-        :param max_wait:
-        :param kwargs:
-        :return:
+        Snake-to-camel cannot handle certain namespaces, but definitions are thankfully case-agnostic.
+        e.g., Namespace: ed-fi-xassessment-roster: (Expected: edFiXassessmentRoster; Actual: edFiXAssessmentRoster)
         """
-        raise NotImplementedError
-
-
-    @abc.abstractmethod
-    def total_count(self) -> int:
+        definitions = {id.lower(): define for id, define in self.swagger.definitions.items()}
+        return definitions.get(self.definition_id.lower(), {})
+    
+    def validate(self, payload: dict):
         """
-        Ed-Fi 3 resources/descriptors can be fed an optional 'totalCount' parameter in GETs.
-        This returns a 'Total-Count' in the response headers that gives the total number of rows for that resource with the specified params.
-        Non-pagination params (i.e., offset and limit) have no impact on the returned total.
-
-        :return:
+        Validate a payload against the expected endpoint structure, as outlined in its Swagger definition.
         """
-        raise NotImplementedError
-
+        # Create the validator only once to improve performance when validating many rows.
+        if not self.validator:
+            self.validator = jsonschema.Draft4Validator(self.definition)
+        self.validator.validate(payload)
 
     @property
-    def description(self):
-        if self._description is None:
-            self._description = self._get_attributes_from_swagger()['description']
-        return self._description
+    def fields(self) -> List[str]:
+        return list(self.field_dtypes.keys())
 
     @property
-    def has_deletes(self):
-        if self._has_deletes is None:
-            self._has_deletes = self._get_attributes_from_swagger()['has_deletes']
-        return self._has_deletes
+    def field_dtypes(self) -> Dict[str, str]:
+        return self._recurse_definition_schema(self.definition)['field_dtypes']
+        
+    @property
+    def required_fields(self) -> List[str]:
+        return list(self._recurse_definition_schema(self.definition)['required'])
 
+    @property
+    def identity_fields(self) -> List[str]:
+        return list(self._recurse_definition_schema(self.definition)['identity'])
 
-    def _get_attributes_from_swagger(self):
-        """
-        Retrieve endpoint-metadata from the Swagger document.
+    @property
+    def has_deletes(self) -> bool:
+        return self.swagger.get_endpoint_deletes().get((self.namespace, self.name))
 
-        Populate the respective swagger object in `self.client` if not already populated.
+    @property
+    def description(self) -> Optional[str]:
+        return self.swagger.get_endpoint_descriptions().get(self.name)
+    
+    def _recurse_definition_schema(self,
```

Comment:
This method needs to be validated for whether its output is both correct and recurses correctly.